extension/registry.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extension
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/pingcap/errors"
+)
+
+type registry struct {
+	sync.RWMutex
+	factories      map[string]func() ([]Option, error)
+	extensionNames []string
+	setup          bool
+	extensions     *Extensions
+	close          func()
+}
+
+// Extensions returns the extensions after setup
+func (r *registry) Extensions() (*Extensions, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if !r.setup {
+		return nil, errors.New("The extensions has not been setup")
+	}
+	return r.extensions, nil
+}
+
+// RegisterFactory registers a new extension with a factory
+func (r *registry) RegisterFactory(name string, factory func() ([]Option, error)) error {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.setup {
+		return errors.New("Cannot register new extension because registry has already been setup")
+	}
+
+	if name == "" {
+		return errors.New("extension name should not be empty")
+	}
+
+	if _, ok := r.factories[name]; ok {
+		return errors.Errorf("extension with name '%s' already registered", name)
+	}
+
+	if r.factories == nil {
+		r.factories = make(map[string]func() ([]Option, error))
+	}
+
+	r.factories[name] = factory
+	r.extensionNames = append(r.extensionNames, name)
+	sort.Strings(r.extensionNames)
+	return nil
+}
+
+// Setup setups all extensions
+func (r *registry) Setup() (err error) {
+	r.Lock()
+	defer r.Unlock()
+	if r.setup {
+		return nil
+	}
+
+	if len(r.factories) == 0 {
+		r.extensions = nil
+		r.setup = true
+		return nil
+	}
+
+	clearBuilder := &clearFuncBuilder{}
+	defer func() {
+		if err != nil {
+			clearBuilder.Build()()
+		}
+	}()
```

Comment:
Isn't this duplicated with `r.close`?

Reply:
If reached line 87, `r.close` is not initialized successfully yet